carbon-design-system/carbon
# Added new React 18 hook `useId` and tests

Closes #11513

It was added the new hook in React 18 called `useId`. This hook will take of hydration for us now. 
We are checking if the user is running a compatible version and calling the function `nativeReactUseId`, otherwise we will keep using our old logic.

#### Changelog

**New**

- Added `useId` from React 18
- Added tests for the `useFallbackId` hook

**Changed**

- Changed the `useId.server-test.js` test. With this new `useId` hook an id will be genarate in the server also.

#### Testing / Reviewing

To test the code you have to switch between React versions. 

I created a [Stackblitz](https://stackblitz.com/edit/github-b5r8er-rfgyun?file=src%2FApp.jsx) where you can test that by following these steps:

- Change the React version to an older version (17, 16...) in the `package.json`
- Restart the server by running `npm install` and `npm run dev` 

In the right panel you should see the changes.

___________

To check in the Storybook you can go to the Accordion component and check if it has an id populate with the `useId` hook

![Screenshot 2023-06-20 at 17 10 04](https://github.com/carbon-design-system/carbon/assets/52183462/16ae76ea-990b-47bc-b03f-a770bd6560c0)
